### PR TITLE
polish(nyz): polish dqn and ppo comments

### PR DESCRIPTION
**Pull request by PaParaZz1**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/policy/r2d2.py [hunk 1]
```diff
@@ -15,16 +15,16 @@
...
 @POLICY_REGISTRY.register('r2d2')
 class R2D2Policy(Policy):
-    r"""
+    """
     Overview:
         Policy class of R2D2, from paper `Recurrent Experience Replay in Distributed Reinforcement Learning` .
-        R2D2 proposes that several tricks should be used to improve upon DRQN,
-        namely some recurrent experience replay tricks such as burn-in.
+        R2D2 proposes that several tricks should be used to improve upon DRQN, namely some recurrent experience replay \
+        tricks and the burn-in mechanism for off-policy training.
```
Comment (puyuan1996) [inline ding/policy/r2d2.py:22]: The R2D2 policy class is inspired by the paper "Recurrent Experience Replay in Distributed Reinforcement Learning". R2D2 suggests the incorporation of several enhancements over DRQN, specifically the application of novel recurrent experience replay strategies and the implementation of a burn-in mechanism for off-policy training.

File: ding/policy/r2d2.py [hunk 2]
```diff
@@ -72,9 +72,6 @@ class R2D2Policy(Policy):
         priority=True,
         # (bool) Whether use Importance Sampling Weight to correct biased update. If True, priority must be True.
         priority_IS_weight=True,
```
Comment (puyuan1996) [inline ding/policy/r2d2.py:74]: Whether to *

File: ding/policy/r2d2.py [hunk 3]
```diff
@@ -85,64 +82,103 @@ class R2D2Policy(Policy):
...
+                # (float) Epsilon start value.
                 start=0.95,
+                # (float) Epsilon end value.
                 end=0.05,
+                # (int) Decay length(env step).
                 decay=10000,
             ),
-            replay_buffer=dict(replay_buffer_size=10000, ),
+            replay_buffer=dict(
+                # (int) Maximum size of replay buffer. Usually, larger buffer size is better.
```
Comment (puyuan1996) [inline ding/policy/r2d2.py:142]: 与之前类似的问题

File: ding/policy/r2d2.py [hunk 4]
```diff
@@ -85,64 +82,103 @@ class R2D2Policy(Policy):
...
+            # (int) `env_num` is used in hidden state, should equal to that one in env config.
             # User should specify this value in user config.
             env_num=None,
         ),
+        # eval_mode config
         eval=dict(
-            # `env_num` is used in hidden state, should equal to that one in env config.
+            # (int) `env_num` is used in hidden state, should equal to that one in env config.
             # User should specify this value in user config.
             env_num=None,
```
Comment (puyuan1996) [inline ding/policy/r2d2.py:127]: 说明None的含义，这里是否改为一个默认值

File: ding/policy/r2d2.py [hunk 5]
```diff
@@ -85,64 +82,103 @@ class R2D2Policy(Policy):
...
+        # learn_mode config
         learn=dict(
+            # (int) How many updates(iterations) to train after collector's one collection.
+            # Bigger "update_per_collect" means bigger off-policy.
+            # collect data -> update policy-> collect data -> ...
             update_per_collect=1,
+            # (int) How many samples in a training batch.
             batch_size=64,
+            # (float) The step size of gradient descent.
             learning_rate=0.0001,
```
Comment (puyuan1996) [inline ding/policy/r2d2.py:94]:
```python
# (int) The number of training updates (iterations) to perform after each data collection by the collector.
# A larger "update_per_collect" value implies a more off-policy approach.
# The data collection and policy update process follows this cycle: collect data -> update policy -> collect data -> ...
update_per_collect=1,
# (int) The number of samples in a training batch.
batch_size=64,
# (float) The step size for gradient descent, determining the rate of learning.
learning_rate=0.0001,
```

File: ding/policy/r2d2.py [hunk 6]
```diff
@@ -85,64 +82,103 @@ class R2D2Policy(Policy):
...
+            # Epsilon greedy with decay.
             eps=dict(
+                # (str) Decay type. Support ['exp', 'linear'].
                 type='exp',
+                # (float) Epsilon start value.
                 start=0.95,
+                # (float) Epsilon end value.
                 end=0.05,
+                # (int) Decay length(env step).
                 decay=10000,
```
Comment (puyuan1996) [inline ding/policy/r2d2.py:139]:
```python
# Configuration for epsilon-greedy strategy with decay.
eps=dict(
    # (str) Type of decay. Supports either 'exp' (exponential) or 'linear'.
    type='exp',
    # (float) Initial value of epsilon at the start.
    start=0.95,
    # (float) Final value of epsilon after decay.
    end=0.05,
    # (int) The number of environment steps over which epsilon should decay.
    decay=10000,
```

File: ding/policy/r2d2.py [hunk 7]
```diff
@@ -404,62 +513,104 @@ def _forward_collect(self, data: dict, eps: float) -> dict:
...
-            - obs (:obj:`Any`): Env observation
-            - model_output (:obj:`dict`): Output of collect model, including at least ['action', 'prev_state']
-            - timestep (:obj:`namedtuple`): Output after env step, including at least ['reward', 'done'] \
-                (here 'obs' indicates obs after env step).
+            - obs (:obj:`torch.Tensor`): The env observation of current timestep, such as stacked 2D image in Atari.
+            - policy_output (:obj:`Dict[str, torch.Tensor]`): The output of the policy network with the observation \
+                as input. For R2D2, it contains the action and the prev_state of RNN.
+            - timestep (:obj:`namedtuple`): The execution result namedtuple returned by the environment step method, \
+                except all the elements have been transformed into tensor data. Usually, it contains the next obs, \
+                reward, done, info, etc.
```
Comment (puyuan1996) [inline ding/policy/r2d2.py:540]: - obs: The environment observation at the current timestep (e.g., a stacked 2D image in Atari).
- policy_output: The output from the policy network given the observation as input. For R2D2, this includes the action and the ``prev_state`` of the RNN.
- timestep: The result returned by the environment's step method, with all elements transformed 
            into tensor data. Usually, this includes the ``next_obs``, ``reward``, ``done``, ``info``(other information).

File: ding/policy/r2d2.py [hunk 8]
```diff
@@ -404,62 +513,104 @@ def _forward_collect(self, data: dict, eps: float) -> dict:
...
 
-    def _process_transition(self, obs: Any, model_output: dict, timestep: namedtuple) -> dict:
-        r"""
+    def _process_transition(self, obs: torch.Tensor, policy_output: Dict[str, torch.Tensor],
+                            timestep: namedtuple) -> Dict[str, torch.Tensor]:
+        """
         Overview:
-            Generate dict type transition data from inputs.
+            Process and pack one timestep transition data info a dict, which can be directly used for training and \
+            saved in replay buffer. For R2D2, it contains obs, action, prev_state, reward, done.
```
Comment (puyuan1996) [inline ding/policy/r2d2.py:533]: Processes and packages a single timestep transition data into a dict. This data can be used directly for training or stored in a replay buffer. For R2D2, the data includes ``obs``, ``action``, ``prev_state``, ``reward``, and ``done``.

File: ding/policy/dqn.py [hunk 9]
```diff
@@ -344,38 +392,40 @@ def _forward_collect(self, data: Dict[int, Any], eps: float) -> Dict[int, Any]:
...
         Overview:
-            For a given trajectory(transitions, a list of transition) data, process it into a list of sample that \
-            can be used for training directly. A train sample can be a processed transition(DQN with nstep TD) \
-            or some continuous transitions(DRQN).
+            For a given trajectory (transitions, a list of transition) data, process it into a list of sample that \
+            can be used for training directly. In DQN with nstep TD, a train sample is a processed transition. \
+            This method is usually used in collectors to execute necessary \
+            RL data preprocessing before training, which can help learner amortize revelant time consumption. \
+            In addition, you can also implement this method as an identity function and do the data processing \
+            in ``self._forward_learn`` method.
```
Comment (puyuan1996) [inline ding/policy/dqn.py:403]: 一些语句或许可以用GPT4润色一下
Processes a given trajectory (a list of transitions) into a list of samples that can be used directly for training. For example, in DQN with n-step TD, a training sample is a processed transition. 

This method is commonly used in collectors to perform necessary pre-processing of RL data before training, 
helping to amortize the time spent on these operations. Alternatively, this method can be implemented as an 
identity function, and data processing can be performed in the `self._forward_learn` method.

File: ding/policy/dqn.py [hunk 10]
```diff
@@ -344,38 +392,40 @@ def _forward_collect(self, data: Dict[int, Any], eps: float) -> Dict[int, Any]:
...
         Arguments:
-            - data (:obj:`List[Dict[str, Any]`): The trajectory data(a list of transition), each element is the same \
-                format as the return value of ``self._process_transition`` method.
+            - transitions (:obj:`List[Dict[str, Any]`): The trajectory data (a list of transition), each element is \
+                the same format as the return value of ``self._process_transition`` method.
         Returns:
```
Comment (puyuan1996) [inline ding/policy/dqn.py:407]: The trajectory data, which is a list of transitions. Each element is **in** the same format as the return value of the `self._process_transition` method.

File: ding/policy/dqn.py [hunk 11]
```diff
@@ -344,38 +392,40 @@ def _forward_collect(self, data: Dict[int, Any], eps: float) -> Dict[int, Any]:
...
-            - samples (:obj:`dict`): The list of training samples.
-
-        .. note::
-            We will vectorize ``process_transition`` and ``get_train_sample`` method in the following release version. \
-            And the user can customize the this data processing procecure by overriding this two methods and collector \
-            itself.
+            - samples (:obj:`List[Dict[str, Any]]`): The processed train samples, each element is the similar format \
+                as input transitions, but may contain more data for training, such as nstep reward and target obs.
         """
```
Comment (puyuan1996) [inline ding/policy/dqn.py:410]: The processed training samples. Each element is similar in format to the input transitions, but may contain additional data for training, such as n-step reward and target observations.

File: ding/policy/base_policy.py [hunk 12]
```diff
@@ -1,22 +1,39 @@
...
 from ding.utils import import_module, allreduce, broadcast, get_rank, allreduce_async, synchronize, deep_merge_dicts, \
     POLICY_REGISTRY
 
 
 class Policy(ABC):
+    """
+    Overview:
+        The basic class of Reinforcement Learning (RL) and Imitation Learning (IL) policy in DI-engine.
+    Property:
+        ``cfg``, ``learn_mode``, ``collect_mode``, ``eval_mode``
```
[the remark this answers is not in the document]
Comment (PaParaZz1) [reply ding/policy/base_policy.py:19]: Property统一为``xxx``

File: ding/policy/base_policy.py [hunk 13]
```diff
@@ -110,13 +156,26 @@ def __init__(
...
+        Arguments:
+            - model (:obj:`torch.nn.Module`): The neural network model to be trained.
+            - bp_update_sync (:obj:`bool`): Whether to synchronize update the model parameters after allreduce the \
+                gradients of model parameters. Async update can be parallel in different network layers like pipeline \
+                so that it can save time.
+        """
         for name, param in model.state_dict().items():
             assert isinstance(param.data, torch.Tensor), type(param.data)
             broadcast(param.data, 0)
+        # here we manually set the gradient to zero tensor at the beginning of the training, which is necessary for
```
[the remark this answers is not in the document]
Comment (PaParaZz1) [reply ding/policy/base_policy.py:177]: 这个无所谓

File: ding/policy/base_policy.py [hunk 14]
```diff
@@ -154,18 +228,77 @@ def cfg(self) -> EasyDict:
...
+        .. note::
+            For the member variables that need to be saved and loaded, please refer to the ``_state_dict_learn`` \
+            and ``_load_state_dict_learn`` methods.
+
+        .. note::
+            For the member variables that need to be monitored, please refer to the ``_monitor_vars_learn`` method.
+
+        .. note::
+            If you want to set some spacial member variables in ``_init_learn`` method, you'd better name them \
+            with prefix ``_learn_`` to avoid conflict with other modes, such as ``self._learn_attr1``.
```
[the remark this answers is not in the document]
Comment (PaParaZz1) [reply ding/policy/base_policy.py:246]: 不用

File: ding/policy/dqn.py [hunk 15]
```diff
@@ -204,23 +219,36 @@ def _init_learn(self) -> None:
...
+            The input value can be torch.Tensor or dict/list combinations and current policy supports all of them. \
+            For the data type that not supported, the main reason is that the corresponding model does not support it. \
+            You can implement you own model rather than use the default model. For more information, please raise an \
+            issue in GitHub repo and we will continue to follow up.
+
+        .. note::
+            For more detailed examples, please refer to our unittest for DQNPolicy: ``ding.policy.tests.test_dqn``.
         """
+        # Data preprocessing operations, such as stack data, cpu to cuda device
         data = default_preprocess_learn(
```
[the remark this answers is not in the document]
Comment (PaParaZz1) [reply ding/policy/dqn.py:252]: 里面的细节注释以后再更新
